src/features/contracts/index.ts

```diff
@@ -0,0 +1,196 @@
+	const repos: RepositoryInfo[] = [
+		{
+			owner: CONTRACTS_PUBLIC_REPO_OWNER,
+			name: CONTRACTS_PUBLIC_REPO_NAME,
+		},
+	];
+
+	if (
+		CONTRACTS_PRIVATE_REPO_OWNER &&
+		CONTRACTS_PRIVATE_REPO_NAME &&
+		CONTRACTS_PRIVATE_REPO_TOKEN
+	) {
+		repos.push({
+			owner: CONTRACTS_PRIVATE_REPO_OWNER,
+			name: CONTRACTS_PRIVATE_REPO_NAME,
+			token: CONTRACTS_PRIVATE_REPO_TOKEN,
+		});
+	}
+
+	return repos;
+};
+
+type FieldsMap = {
+	[dbField: string]: {
+		contractField: string;
+		default?: any;
+	};
+};
+// This map will hold information on which contract fields imported from the contract type will be synced to which db model and fields.
+type SyncSettings = {
+	[contractType: string]: {
+		resource: string;
+		uniqueKey: string;
+		map: FieldsMap;
+	};
+};
+
+let globalSyncSettings: SyncSettings | undefined;
+
+export function setSyncSettings(syncSettings: SyncSettings) {
+	globalSyncSettings = syncSettings;
+}
+
+const mapModel = (contractEntry: Contract, map: FieldsMap) => {
+	const mappedModel: { [k: string]: any } = {};
+	for (const key of Object.keys(map) as Array<keyof typeof map>) {
+		mappedModel[key] =
+			_.get(contractEntry, map[key]?.contractField) ?? map[key]!.default;
+	}
+
+	return mappedModel;
+};
+
+const upsertEntries = async (
+	newData: any[],
+	existingData: Set<string | number | boolean>,
+	resource: string,
+	uniqueField: string,
+	rootApi: sbvrUtils.PinejsClient,
+) => {
+	await Bluebird.map(
+		newData,
+		async (entry: any) => {
+			try {
+				if (existingData.has(entry[uniqueField])) {
+					return await rootApi.patch({
+						resource,
+						body: entry,
+						options: {
+							$filter: { slug: entry[uniqueField] },
+							returnResource: false,
+						},
+					});
+				}
+
+				await rootApi.post({
+					resource,
+					body: entry,
+					options: { returnResource: false },
+				});
+			} catch (err) {
+				console.error(
+					`Failed to synchronize ${entry[uniqueField]}, skipping...`,
+					err.message,
+				);
+			}
+		},
+		{ concurrency: 10 },
+	);
+};
+
+const syncContractsToDb = async (
+	type: string,
+	contracts: Contract[],
+	syncSettings: SyncSettings | undefined,
+) => {
+	const typeMap = syncSettings?.[type];
+	if (!typeMap) {
+		throw new Error(`Contract does not have a corresponding mapping: ${type}`);
+	}
+
+	const mappedModel = contracts.map((contract) =>
+		mapModel(contract, typeMap.map),
+	);
+
+	const rootApi = sbvrUtils.api.resin.clone({
+		passthrough: { req: permissions.root },
+	});
+
+	const existingEntries = (await rootApi.get({
+		resource: typeMap.resource,
+		options: { $select: [typeMap.uniqueKey] },
+	})) as Array<{ [k in typeof typeMap.uniqueKey]: any }>;
+
+	const existingKeys = new Set(
+		existingEntries.map((existingEntry) => existingEntry[typeMap.uniqueKey]),
+	);
+
+	await upsertEntries(
+		mappedModel,
+		existingKeys,
+		typeMap.resource,
+		typeMap.uniqueKey,
+		rootApi,
+	);
+};
+
+export const synchronizeContracts = memoizee(
```

Comment:
Is there any reason to memoize this? It should only be called in a single place when in runtime code so I'm not sure what the intention of memoizing is

Reply:
@Page- It was simply done so we can await the same promise in the tests, but I guess it's fine to just call synchronizeContracts twice, it shouldn't matter. let me remove it.